### PR TITLE
Feature/179733229 marker genes heatmap show cell types

### PR DESCRIPTION
**Pull request by lingyun1010**
I added json endpoints for cell type marker gene heatmap data  and cell type options.

### CODE REVIEW

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java [hunk 1]
```diff
@@ -29,16 +30,26 @@ public MarkerGeneService(MarkerGenesDao markerGenesDao, CellTypeSearchDao cellTy
         var ontologyLabelsCellTypeValues = cellTypeSearchDao.getInferredCellTypeOntologyLabels(experimentAccession, organismPart);
 
         var cellTypeMarkerGenes = ontologyLabelsCellTypeValues.isEmpty() ?
-                markerGenesDao.getCellTypeMarkerGenes(experimentAccession,
+                markerGenesDao.getCellTypeMarkerGenes(experimentAccession, CELL_GROUP_TYPE,
```
Comment (upendrakumbham) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java:35]: @lingyun1010  I'm confused here - why we are passing "inferred cell type - ontology labels" group type for the authors label cell type fetch method. Can you clarify for me?

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java [hunk 2]
```diff
@@ -29,16 +30,26 @@ public MarkerGeneService(MarkerGenesDao markerGenesDao, CellTypeSearchDao cellTy
...
 
         return cellTypeMarkerGenes.stream()
                 .filter(markerGene -> !markerGene.cellGroupValue().equalsIgnoreCase("Not available"))
                 .filter(markerGene -> !markerGene.cellGroupValueWhereMarker().equalsIgnoreCase("Not available"))
                 .collect(toImmutableList());
     }
 
+    public ImmutableList<MarkerGene> getCellTypeMarkerGeneHeatmap(String experimentAccession, String cellGroupType, ImmutableSet<String> cellType) {
+
+
```
Comment (upendrakumbham) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java:51]: I think extra white space here..

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGenesDaoIT.java [hunk 3]
```diff
@@ -111,4 +111,18 @@ void shouldFetchOnlyInferredCellTypeMarkerGenes() {
...
+
+    @Test
+    void shouldGetCellTypesWithMarkerGenes() {
+        var cellTypesWithMarkerGenes = subject.getCellTypesWithMarkerGenes("E-MTAB-5061", "Inferred cell type - ontology labels");
+        assertThat(cellTypesWithMarkerGenes).isNotEmpty();
+
+    }
+
+    @Test
+    void shoulGgetCellTypeMarkerGenes() {
```
Comment (upendrakumbham) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGenesDaoIT.java:123]: - Typo mistake. 
- The above test method and current test method names are identical, but we are testing different functionality. Can you elaborate on test method names to improve readability?

**Comment by upendrakumbham** [comment]
Just adding here to remember the points that we were discussed in the meeting before we forget :) while fixing.

- Fetch cell types for the heatmap by giving input as a cell group type(ontology or authors labels)
     endpoint - `/json/experiments/{experimentAccession}/marker-genes-heatmap/cell-types`
     method- `getCellTypesForMarkerGenesHeatmap`

- Fetch marker genes heatmap data by passing above endpoint resulted cell type as a input
 endpoint- `/json/experiments/{experimentAccession}/marker-genes-heatmap/cell-types` 
 method- `getCellTypeMarkerGeneHeatmapData`
Please change accordingly in the `service` and `dao` layers for both endpoints including tests.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java (hunk 1 above)
Comment (lingyun1010) [reply app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java:35]: This is a refactoring, `inferred cell type -ontology labels` was hard coded in DAO class. To reuse that DAO methods, for general `CELL_GROUP_TYPE`, I did the refactoring by moving the hard-coded value in the service class.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGenesDaoIT.java (hunk 3 above)
Comment (lingyun1010) [reply app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGenesDaoIT.java]: Yes, I refactored here.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java (hunk 2 above)
Comment (lingyun1010) [reply app/src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneService.java:48]: Removed it.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGenesDaoIT.java [hunk 4]
```diff
@@ -111,4 +111,18 @@ void shouldFetchOnlyInferredCellTypeMarkerGenes() {
...
+
+    @Test
+    void shouldGetCellTypesWithMarkerGenesGivenCellGroup() {
+        var cellTypesWithMarkerGenes = subject.getCellTypesWithMarkerGenes("E-MTAB-5061", "Inferred cell type - ontology labels");
+        assertThat(cellTypesWithMarkerGenes).isNotEmpty();
+
+    }
+
+    @Test
+    void shouldGetCellTypeMarkerGenesHeatmapData() {
```
Comment (upendrakumbham) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGenesDaoIT.java:123]: test method name would be some thing like this for the readability- `shouldGetMarkerGenesHeatmapDataForTheGivenCellGroupAndCellType()`

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java [hunk 5]
```diff
@@ -43,4 +43,23 @@ public String getCellTypeMarkerGenes(@PathVariable String experimentAccession,
...
+    }
+
+    @GetMapping(value = "/json/experiments/{experimentAccession}/marker-genes-heatmap/cellTypeGroups",
+            produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public String getCellTypeGroupsMarkerGenesHeatmap(@PathVariable String experimentAccession,
+                                                      @RequestParam String cellGroupType) {
+        return GSON.toJson(
+                markerGeneService.getCellTypesWithMarkerGenes(experimentAccession, cellGroupType)
+        );
+    }
```
Comment (upendrakumbham) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java:64]: Fetch cell types for the heatmap by giving input as a cell group type(ontology or authors labels) & experiment accession
endpoint -` /json/experiments/{experimentAccession}/marker-genes-heatmap/cell-types`
method- `getCellTypesForMarkerGenesHeatmap`

Probably you missed our summary of the meeting points :)

Comment (lingyun1010) [reply app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java:64]: For the endpoint `/json/experiments/{experimentAccession}/marker-genes-heatmap/cell-types`, which retrieve the heatmap data, we need to pass both `cellTypeGroup` and `cellTypes`.

I renamed the endpoint and method to clarify we are retrieving `heatmapData`, as `getCellTypeMarkerGenesHeatmapData` and `getCellTypeMarkerGeneHeatmapData` respetively.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java [hunk 6]
```diff
@@ -43,4 +43,23 @@ public String getCellTypeMarkerGenes(@PathVariable String experimentAccession,
                 markerGeneService.getCellTypeMarkerGeneProfile(experimentAccession, ImmutableSet.copyOf(organismPart))
         ));
     }
+
+    @GetMapping(value = "/json/experiments/{experimentAccession}/marker-genes-heatmap/cell-types",
```
Comment (upendrakumbham) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java:47]: Hi @lingyun1010, can you add `WIT` tests for these new endpoints?

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java [hunk 7]
```diff
@@ -43,4 +43,23 @@ public String getCellTypeMarkerGenes(@PathVariable String experimentAccession,
...
+            produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public String getCellTypeMarkerGenesHeatmapData(@PathVariable String experimentAccession,
+                                                @RequestParam String cellGroupType,
+                                                @RequestParam Set<String> cellType) {
+        return GSON.toJson(highchartsHeatmapAdapter.getMarkerGeneHeatmapDataSortedLexicographically(
+                markerGeneService.getCellTypeMarkerGeneHeatmapData(experimentAccession, cellGroupType, ImmutableSet.copyOf(cellType))
+        ));
+    }
+
+    @GetMapping(value = "/json/experiments/{experimentAccession}/marker-genes-heatmap/cellTypeGroups",
```
Comment (upendrakumbham) [inline app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java:57]: same here

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java (hunk 6 above)
Comment (lingyun1010) [reply app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java:47]: We don't have any `WIT` test in sc repo. I think we can go without it.

File: app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java (hunk 7 above)
Comment (upendrakumbham) [reply app/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonMarkerGenesController.java:57]: Add tests here @lingyun1010 - `JsonMarkerGenesControllerWIT `

**Review by upendrakumbham — DISMISSED** [review]
All looks good to me. Please merge away.

**Pull request merged**